### ISSUE TITLE
Thermal Sensing

### ISSUE THREAD

**Issue by Joedang** [psas/lv3.0-airframe#14]
# Thermal sensing for LV3. 

### Description
The temperatures that the airframe will experience -- namely at the motor mount and the tip of the nose cone -- are a bit of an open question. It would be a good idea to have temperature measurements at these points, and perhaps a few other stations along the length of the rocket. 

### Difficulty / Specializations
This would be a good project for a sophomore or junior MME and ECE to collaborate on. It involves data collection, calibration, simple circuit design, simple mechanical design, and collaboration with other projects to make sure you aren't interfering with them. 

##### Must
* Not interfere with any other subsystems (!).
* Record temperatures up to 200 C (400 F).
* Record at a rate of 0.5 Hz or better.
* Record at a resolution of +/- 3 C or better.
* Compensate for the temperature of the board.
* Record the temperature of the thrust flange.
* Not interfere with the assembly/disassembly of the rocket.
* Not catch on fire.

##### Should
* Be able to record temps at 2 to 7 different points throughout the airframe (thrust flange and nose cone tip take priority).
* Record temperatures below whatever the typical temperature is at 8 km above Brother's Oregon.
* Record at a rate of 10 Hz or better.
* Record at a resolution of +/- 0.25 C or better.
* Use the same power, voltage, and data standards that the rest of the flight computer uses, in case we want to integrate it into another board at some point.
* Not add significant complexity to the assembly of the flight computer, recovery system (!), eNSR, or the rocket in general.
* Be easy and cheap to manufacture and test/calibrate.
* Be extensible to use other physical sensors. (This requires the ability to see the future.)
* Be applicable to thermal sensing on OreSat.
* Not shake apart when launched on a rocket. 

### Suggested Action
There are a few ways to go about this, but I think the best right now would be to have a centralized thermal sensing board in the flight computer with cables running to thermocouples throughout the rocket. 

In lower temperature applications, it would make a lot more sense to use a pre-fab thermistor, but those generally don't operate above 130 C (300 F). I'm concerned about whether or not the temperature is exceeding 177 C (350 F), which is the maximum working temperature of the epoxy matrix in the airframe. 

On the electrical side of things, you *could* make high-input-impedance amplifiers for each thermocouple and do the cold-junction compensation and calibration on your own... or you could *not* re-invent the wheel and just use an existing IC to do it. (An initial internet search suggests the [MAX31855](http://datasheets.maximintegrated.com/en/ds/MAX31855.pdf) might be a good choice.) Either way, you have to be able to record the temperature data to and SD card. Whether or not you do this on the thermal sensing board or by sending the data to the flight computer is up to you and the other relevant ECE peeps. You will also need to make sure your design can handle any of the thermocouples being suddenly disconnected.

For the mechanical stuff, the big challenge is figuring out how you're going to pass the thermocouple wires between the modules. You will need something that can disconnect at each coupling ring. Most importantly, you will need something that can reliably disconnect at the eNSR without snagging and without interfering with the recovery system. I was thinking ethernet plugs would work well at the coupling rings. Perhaps the connection at the eNSR could have an ethernet plug with the tap cut off, so it simply pulls apart. Reusing the design from the umbilical disconnect may be a better option though, since it's a flight-tested design. 

##### The electrical team member should:
* Look at possible ICs that could make this easier.
* Talk to people working on the flight computer to figure out whether you can share power, data, and/or board space with them. 
* Figure out what your design constraints are. Where can you run cables in/out of the flight computer? Are you storing the data on the thermal sensing board or on the flight computer? Will you transmit the data to the ground station or just let it be retrieved from an SD card?

##### The mechanical team member should:
* Begin playing with different possible disconnects.
* Talk to people working on the flight computer, recovery (!), eNSR (!), RCS, and airframe to figure out what your design constraints are, so you don't break their projects and vice versa.
* Think about how you will calibrate each thermocouple once it is fixed in place on the rocket.

### Extended Background
Composites work by having lots of strong/rigid fibers gooped up in a flexible matrix. The matrix transfers stress between the fibers. And, because the fibers are very long compared to their width, the matrix has a lot of area with which to transfer this stress. So, the matrix only experiences a small fraction of the total stress and the whole material takes on the properties of the individual fibers. 

If the matrix gets donked up though, it kind of ruins the whole thing. So, it would be bad if our matrix got too hot. It would start to chemically break down and become weaker. We definitely want to know if this is happening. (Plus, getting to experiment with heat transfer and trans/supersonic heating is awesome.) 

A thermocouple is just two different metals welded together. The different metals have valence electrons with different energy levels. This essentially makes a waterfall of electrons from one metal to the other. The hotter this junctions gets, the more often electrons will wander into it. So, higher temperature means higher voltage. To form a circuit you need two such junctions (intermediate junctions just cancel), so if you know the voltage difference, you also need to know the temperature of one junction to know the temperature of the remaining junction. The junction you're measuring is usually referred to as the hot junction and your voltage measuring device is the cold junction (hence "cold junction compensation"). The junctions creates a constant voltage, but it's small (microvolts) and will saturate easily so you need a high gain low input impedance amplifier. Check out the Wikipedia pages on [the Thermoelectric Effect](https://en.wikipedia.org/wiki/Thermoelectric_effect) and [thermocouples](https://en.wikipedia.org/wiki/Thermocouple) if you want to know more.

**Comment by kwilsonpdx** [comment]
* If the system controller board from OreSat is used for lv3.0 avionics, then this issue may be related:

https://github.com/oresat/system-controller/issues/10

**Comment by Joedang** [comment]
Yeah, it sounds like we want to do the same thing. Do you know if anyone is currently working on that? 

Have you guys thought about what connectors would be used on the board? I imagine the standard K-type connectors, for example, would be excessively large.

**Comment by kwilsonpdx** [comment]
* The oresat issue referenced is a proposal for the next system controller board revision.
* Connector type(s) will depend on sensor type(s) (RTD or Thermocouple) and use case (flight or prototype). We should have more insight after we capture the requirements for the next revision.

**Comment by Joedang** [comment]
# Update
After talking with @andrewgreenberg and Kenny, it seems best to have a simple/hacky solution for L-13. Once we are trying to use OreSat as our flight computer, then we can worry about more complicated solutions.

Basically, the sensor should be a small Arduino-compatible board with some off-the-shelf thermocouple and/or thermistor, a LiPo battery that will last at least 6 hours, and an SD card to store its data. The sensors will just be attached to the aluminum bits using thermal epoxy. 

Ideally, it should also have some way of identifying when launch occurs to within 1 second. Ideally, it should append the data to the SD card every second, in case the battery shakes loose.

Whoever puts this together, it's your choice on whether you want to have one board per location or run wires along the length of the rocket. If you go with the latter option, it needs to be very thin wire with lots of slack (~2 ft) near the eNSR, so it will just tear away when the nosecone separates.

**Comment by RocketmanYG** [comment]
Thermal sensing for LV3 is done! We can now store temperature data to an SD card and see how hot LV3 carbon fiber airframe is going to get.

**Comment by Joedang** [comment]
I'm gonna say this can be closed pending three things:  
- [ ] The cad for the mounting/enclosure should be added to this repo. (I know it was done in Onshape, so we can't really put the "source" CAD here. But IIRC, Onshape will export to SLDPRT in a sane way. If not, just create a STEP.) @SaturnVF1 
- [ ] A bill of materials should appear alongside the code and CAD. People who have never seen this before should be able to find the parts online, just based on what's in the BoM. (So, it *must* include manufacturer part numbers and *should* include purchasing links.) 
- [ ] The files should be sorted and named appropriately. The code and CAD should get their own directory under `tools/`, while the data  and its analysis script should go under `test/`. *Also, holy guacamole, why is the analysis script committed but not the data?!?!* @RocketmanYG  I'm assuming the sensor code is `.c` and the analysis is `.m`. Neither file name should include spaces.
